nemo/collections/tts/data/datalayers.py

```diff
@@ -207,3 +217,223 @@ def __getitem__(self, index):
+
+    @torch.no_grad()
+    def collate_fn(self, batch):
+        """ return data with zero-padding
+
+        Important data like x, y are all converted to Tensor(cpu).
+        :param batch:
+        :return: DataDict
+            Values can be an Tensor(cpu), list of str, ndarray of int.
+        """
+
+        result = dict()
+        T_xs = np.array([item.pop('T_x') for item in batch])
+        idxs_sorted = np.argsort(T_xs)
+        T_xs = T_xs[idxs_sorted].tolist()
+        T_ys = [batch[idx].pop('T_y') for idx in idxs_sorted]
+        length = [batch[idx].pop('length') for idx in idxs_sorted]
+
+        result['T_xs'], result['T_ys'], result['length'] = T_xs, T_ys, length
+
+        for key, value in batch[0].items():
+            if type(value) == str:
+                list_data = [batch[idx][key] for idx in idxs_sorted]
+                set_data = set(list_data)
+                if len(set_data) == 1:
+                    result[key] = set_data.pop()
+                else:
+                    result[key] = list_data
+            else:
+                if len(batch) > 1:
+                    # B, T, F, C
+                    data = [batch[idx][key].permute(1, 0, 2) for idx in idxs_sorted]
+                    data = pad_sequence(data, batch_first=True)
+                    # B, C, F, T
+                    data = data.permute(0, 3, 2, 1)
+                else:  # B, C, F, T
+                    data = batch[0][key].unsqueeze(0).permute(0, 3, 1, 2)
+
+                result[key] = data.contiguous()
+
+        x = result['x']
+        mag = result['y_mag']
+        max_length = max(result['length'])
+        y = result['y']
+        T_ys = result['T_ys']
+        length = result['length']
+        path_speech = result['path_speech']
+        return x, mag, max_length, y, T_ys, length, path_speech
+
+        return result
+
+    @staticmethod
+    @torch.no_grad()
+    def decollate_padded(batch: DataDict, idx: int) -> DataDict:
+        """ select the `idx`-th data, get rid of padded zeros and return it.
+
+        Important data like x, y are all converted to ndarray.
+        :param batch:
+        :param idx:
+        :return: DataDict
+            Values can be an str or ndarray.
+        """
+        result = dict()
+        for key, value in batch.items():
+            if type(value) == str:
+                result[key] = value
+            elif type(value) == list:
+                result[key] = value[idx]
+            elif not key.startswith('T_'):
+                T_xy = 'T_xs' if 'x' in key else 'T_ys'
+                value = value[idx, :, :, : batch[T_xy][idx]]  # C, F, T
+                value = value.permute(1, 2, 0).contiguous()  # F, T, C
+                value = value.numpy()
+                if value.shape[-1] == 2:
+                    value = value.view(dtype=np.complex64)  # F, T, 1
+                result[key] = value
+
+        return result
+
+
+def setup(files_list, num_snr, kwargs_stft, dest, desc):

```

Comment:
~What does this function do?~
Nvm, I see what this does. It looks like it is injecting random noise. Is there any reason you don't use the NoisePertrubation that exists in the nemo_asr library and do it on the fly rather than preprocess it?

Reply:
Yeah, I tried it and saw that adding different noise at every step results worse training. I don't have a non-empirical reason for it, but I guess it have something to do with destablizing training on the initial epochs.